docs/specifications/4-standard-compliance-suite.md

```diff
@@ -0,0 +1,166 @@
+  changing the tests case's semantic value) should not be included in the list.
+
+After the attributes table, a **TypeScript** block of code describes the
+canonical form of the test. It includes any type declaration that is used by the
+test (so the code example is self-contained). Assertions performed by the test
+should be written in the form of [`jest`] expectations.
+
+[`jest`]: https://jestjs.io/docs/en/getting-started
+
+Finally, another code block details the sequence of messages that should be
+exchanged between the *host* and `node` processes during the execution of the
+test case, such that implementations can assert conherent behavior. The dialogue
+is the sequence of JSON formatted messages, from the perspective of the *host*
+app, using the following notation:
+* Messages sent by the *host* runtime to the `node` process:
+  ```
+  > { "api": "foo" }
+  ```
+* Messages received by the *host* runtime from the `node` process:
+  ```
+  < { "result": "bar" }
+  ```
+* Comments to improve readability of the trace:
+  ```
+  # Comment continues until the end of the line
+  ```
+* Blank lines can be added to logically group trace elements
+
+> :question: is there a need to support some form of a capture mechanism to
+> provision for non-deterministric results, or non-normative elements such as
+> the exact Object IDs issued for created instances?
+
+#### Template
+
+Below is the template markdown to copy-paste when introducing a new test case in
+the compliance suite. New tests should always be added at the very end of the
+category they belong to, right after the last test in said category.
+
+````md
+### Test Category
+#### Test Case Name
+
+A short english language description of what property this test verifies. The
+description should include enough detail for a reader to be able to understand
+the test without having to search for any additional information. Prefer a long,
+unambiguous description to a terse one that could be subject to interpretation.
+
+Attribute      | Value
+---------------|----------------------------------------------------------------
+Introduced     | 1.2.3
+Related Issues | #123, aws/aws-cdk#456
+
+##### Reference Implementation
+```ts
+// Declare any type used by the test
+export class Foo { /* ... */ }
+
+// Instructions the test will execute, using `jest`-like expectation syntax
+expect(new Foo().bar().baz)
+  .toBeUndefined();
+```
+
+##### Reference Kernel Messaging
+```
+< { "hello": "@jsii/runtime@1.2.3" }
+> { "load": { "name": "test-case-001", "version": "1.2.3", "tarball": "/tmp/jsii-kernel-test/lib.tgz" } }
+< { "assembly": "test-case-001", "types": 3 }
+```
+````
+
+## Compliance
+In order to be able to assert compliance of language binding libraries to the
+standard test suite, implementations are responsible for provide a test harness
+that can produce a standardized test report in the form of a JSON document that
+follows the following schema:
```

Comment:
I'm a bit lost in here. Where would this test harness be checked in? What are other requirements of this harness; what is its entry point and exit points? It feels like we've not fully defined the boundary here and the interactions.

Based on my reading so far, it seems that there is a compliance test runner somewhere that interacts with the harness which further interacts with test categories and cases. It's not clear to me what these interactions and components are. Are these going to be clarified elsewhere? 
On the other hand, is this a bit heavy handed for the first round of compliance tests? Anything simpler we can do here?

Reply:
My perspective is each implementation defines it's own test suite & harness. The contract is it must produce the correct output document (as specified here); and a CLI will be able to consume that output document + the test suite markdown document; and and provides an assessment of coverage (%Compliance, or otherwise) -- this is then made part of the CI/CD pipeline.